R/nimble_lm.R

```diff
@@ -219,64 +399,103 @@ nimble.lm <- function(mod, dat=NULL, family=c("Normal"), priors=c("Normal","t","
   family = match.arg(family)
   prior = match.arg(priors)
 
-  mod = as.formula(mod)
 
   index.name <- quote(i)
 
   #Likelihood
+
+  #Generate mean function mu[i]
   if (family=="Normal"){
     meanfcn <- quote(dnorm(mu[i],tau))}  #could change to other disributions later on
-    dep <- LHS2BUGSterm(mod[[2]], index.name)
+    dep <- LHS2BUGSterm(mf$formula[[2]], index.name)
     meanfctn <-  substitute(LHS ~ RHS, list(LHS = dep, RHS = meanfcn))
 
+    #Pull out Fixed Terms
+    fixedTm <- nobars(formula)
+    terms.to.add <- attr(terms(fixedTm), 'term.labels')
 
-    #Generate mean function mu[i]
-    #Intercept
-    int.to.add <- attr(terms(mod), 'intercept')
+    randomTerms <- fb(formula)  #Puts Random Effect Terms in a List
 
-    terms.to.add <- attr(terms(mod), 'term.labels')
-    assert_that(length(terms.to.add) > 0)
+    #Store Random Effect Terms and Priors
+    RandomParms <- list()
+    RandomPriors <- list()
+
+    #Generate Random Terms and Priors First
+    if(!is.null(randomTerms)){
+      for (i in 1:length(randomTerms)){
+        RandomParms[[i]] <- formulaRandom2BUGS(randomTerms[[i]], fr, index.name, intcpt = int.add, fixedterms = terms.to.add)[[1]]
+        RandomPriors[[i]] <- formulaRandom2BUGS(randomTerms[[i]], fr, index.name, intcpt = int.add, fixedterms = terms.to.add)[[2]]
+      }
+    }
+
+  #Add Fixed Effect Terms
+  assert_that(length(terms.to.add) > 0 | length(randomTerms) > 0)
 
 
-  ## Assume Fixed Intercept for Now
-  if (int.to.add==1){
+
+  ## Check for Overal Model Intercept
+  if (int.add==1){
       RHS <- quote(b0)
       ## append additional terms
       for(iTerm in seq_along(terms.to.add)) {
-        RHS <- addTerm(RHS, formulaTerm2BUGSterm(terms.to.add[iTerm], mf, index.name,id = iTerm))
+        RHS <- addTerm(RHS, formulaTerm2BUGSterm(terms.to.add[iTerm], fr, index.name, id = iTerm))
     }
   }
 
-  if(int.to.add==0){
-    RHS <- formulaTerm2BUGSterm(terms.to.add[1], mf, index.name, id=1)
-    for(iTerm in seq_along(terms.to.add[-1])+1) {
-      RHS <- addTerm(RHS, formulaTerm2BUGSterm(terms.to.add[iTerm], mf, index.name,id = iTerm))
+  if(int.add==0 & length(terms.to.add) > 0){
+    RHS <- formulaTerm2BUGSterm(terms.to.add[1], fr, index.name, id = 1)
+    for(iTerm in seq_along(terms.to.add[-1]) + 1){
+      RHS <- addTerm(RHS, formulaTerm2BUGSterm(terms.to.add[iTerm], fr, index.name, id = iTerm))
     }
   }
 
 
+  #Add Back in Random Terms; Need Check if There are No Fixed Terms Or Intercept Though to Define RHS
+  if(!is.null(randomTerms) & exists("RHS")){
+    for (i in 1:length(RandomParms)){
+      RHS <- addTerm(RHS, RandomParms[[i]])
+    }
+  }
+
+  if(!is.null(randomTerms) & !exists("RHS")){
+    RHS <- RandomParms[[1]]
+    for (i in seq_along(RandomParms[-1]) + 1){
+      RHS <- addTerm(RHS, RandomParms[[i]])
+    }
+  }
+
   #Make Term for mean function, mu
   LHS <- LHS2BUGSterm(quote(mu), index.name)
   fullLine <- substitute(LHS ~ RHS, list(LHS = LHS, RHS = RHS))
```

Comment:
Should this use `<-` instead of  `~`.

Also I suggest a name like `mu` should be generated based on the name of the variable like `Y`, e.g. `Y.predicted`.  That way we can offer user control over the name generation and we can avoid problems if a user has a variable called `mu`, e.g. `nimble.lm( Y ~ A + (1 | mu) )`